### PR TITLE
Build System Improvement - Part 1 : main build.gradle file 

### PR DESCRIPTION
**Pull request by mejdoo**
## What's new in this PR?

This PR comes with a refactoring solution to improve the readability the current build system which is very complicated and hard to understand.

## Main improvements
- Move the gradle tasks and configurations to different scripts
- Manage the versioning from a single place : `Dependencies.kt`

#### APK
[Download build #2912](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2912/artifact/build/artifact/wire-dev-PR3073-2912.apk)
[Download build #2916](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2916/artifact/build/artifact/wire-dev-PR3073-2916.apk)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -33,258 +26,29 @@ plugins {
...
+        maven {url "https://oss.sonatype.org/content/repositories/releases"}
+        maven {url "https://oss.sonatype.org/content/repositories/snapshots"}
     }
 }
 
-apply from: "./scripts/detekt.gradle"
-apply from: "./scripts/jacoco.gradle"
-apply from: "./scripts/avs.gradle"
-
-// FIXME: License header is disabled because it breaks compilation
```
Comment (mejdoo) [inline build.gradle:59]: I will try to solve this in future PRs. I coudn't find a quick solution.

File: scripts/license.gradle [hunk 2]
```diff
@@ -0,0 +1,23 @@
+license {
+    header = file('LICENSE_HEADER')
+    ext.year = Calendar.getInstance().get(Calendar.YEAR)
+    ext.name = 'Wire Swiss GmbH'
+    skipExistingHeaders = true
+}
+
+//TODO fix compilation issue for the license plugin
```
Comment (mejdoo) [inline scripts/license.gradle:8]: I will try to solve this in future PRs. I coudn't find a quick solution.

**Review by android10 — APPROVED** [review]
:+1:

File: zmessaging/build.gradle [hunk 3]
```diff
@@ -1,17 +1,17 @@
 apply plugin: 'com.android.library'
-apply plugin: 'kotlin-android-extensions'
 apply plugin: 'kotlin-android'
+apply plugin: 'kotlin-android-extensions'
```
Comment (mejdoo) [inline zmessaging/build.gradle:3]: @gizemb here is the fix for the warning.

```
> Configure project :zmessaging
Kotlin plugin should be enabled before 'kotlin-android-extensions'
```

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 4]
```diff
@@ -1,17 +1,32 @@
...
 
     //core
     const val KOTLIN = "1.3.72"
+
     const val WIRE_TRANSLATIONS = "1.+"
     val WIRE_AUDIO = System.getenv("AUDIO_VERSION") ?: "1.209.0@aar"
 
     //plugins
+    const val GRADLE = "3.2.1"
+    const val SCALA = "1.6"
```
Comment (gizemb) [inline buildSrc/src/main/kotlin/Dependencies.kt:24]: `const val SCALA_BUILD_PLUGIN`

the version of the Scala language that we use is 2.12 and `SCALA = 1.6` seems misleading here

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 5]
```diff
@@ -1,17 +1,32 @@
...
+    val TARGET_COMPATIBILITY_VERSION = JavaVersion.VERSION_1_8
 
     //core
     const val KOTLIN = "1.3.72"
+
     const val WIRE_TRANSLATIONS = "1.+"
     val WIRE_AUDIO = System.getenv("AUDIO_VERSION") ?: "1.209.0@aar"
 
     //plugins
+    const val GRADLE = "3.2.1"
```
Comment (gizemb) [inline buildSrc/src/main/kotlin/Dependencies.kt:23]: `const val ANDROID_GRADLE_PLUGIN`

build plugin version and Gradle version itself are separate things https://developer.android.com/studio/releases/gradle-plugin#updating-gradle

File: scripts/ci.gradle [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+class CheckstyleIssue extends Issue {
+    CheckstyleIssue(File file, Node node) {
+        this.file = file
+        line = node.'@line'
+        column = node.'@column'
+        error = node.'@message'
+    }
+}
+
+task clean(type: Delete) {
```
Comment (gizemb) [inline scripts/ci.gradle:119]: this is not necessarily related to ci

File: zmessaging/build.gradle [hunk 7]
```diff
@@ -43,22 +43,6 @@ android {
...
-    google()
-    maven {
-        url "https://oss.sonatype.org/content/repositories/releases"
-    }
-    maven {
-        url "https://oss.sonatype.org/content/repositories/snapshots"
-    }
-    maven { url "https://dl.bintray.com/wire-android/releases/" }
-    maven { url "https://dl.bintray.com/wire-android/snapshots/" }
-    maven { url "https://dl.bintray.com/wire-android/third-party/" }
```
Comment (gizemb) [inline zmessaging/build.gradle:58]: if these repos are only needed for this module, I think it's better to keep them here. if you apply this to `allProjects`, then let's say `app` module will go through these directories too when looking for its dependencies, which might add up to total build time.

**Review by gizemb — CHANGES_REQUESTED** [review]
I love how organized this looks! just had a couple of comments

File: build.gradle [hunk 8]
```diff
@@ -33,258 +26,29 @@ plugins {
...
-        column = node.'@column'
-        error = node.'@message'
-    }
-}
\ No newline at end of file
+apply from: "./scripts/avs.gradle"
+apply from: "./scripts/detekt.gradle"
+apply from: "./scripts/jacoco.gradle"
+apply from: "./scripts/license.gradle"
+apply from: "./scripts/customization.gradle"
```
Comment (BradleyWire) [inline build.gradle:53]: What is customisation?

File: build.gradle [hunk 9]
```diff
@@ -33,258 +26,29 @@ plugins {
...
-    def CheckstyleIssue(File file, Node node) {
-        this.file = file
-        line = node.'@line'
-        column = node.'@column'
-        error = node.'@message'
-    }
-}
\ No newline at end of file
+apply from: "./scripts/avs.gradle"
+apply from: "./scripts/detekt.gradle"
```
Comment (BradleyWire) [inline build.gradle:50]: `detekt` and `jacoco` could in theory be merged into one script called quality

File: scripts/ci.gradle (hunk 6 above)
Comment (mejdoo) [reply scripts/ci.gradle:119]: where do you suggest to move this?

File: scripts/customization.gradle [hunk 10]
```diff
@@ -0,0 +1,88 @@
+import groovy.json.JsonOutput
```
Comment (BradleyWire) [inline scripts/customization.gradle:1]: I don't think customisation is meaningful enough name to tell us what it does, is there not a terminology wire uses for this process at all?

File: scripts/license.gradle [hunk 11]
```diff
@@ -0,0 +1,23 @@
+license {
```
Comment (BradleyWire) [inline scripts/license.gradle:1]: the name would read better as `licensing.gradle`, this reads that this is a license

File: custom-checkstyle/build.gradle [hunk 12]
```diff
@@ -1,7 +1,7 @@
 apply plugin: 'java'
 
-sourceCompatibility = rootProject.ext.sourceCompatibilityVersion
-targetCompatibility = rootProject.ext.targetCompatibilityVersion
+sourceCompatibility = Versions.SOURCE_COMPATIBILITY_VERSION
+targetCompatibility =  Versions.TARGET_COMPATIBILITY_VERSION
```
Comment (BradleyWire) [inline custom-checkstyle/build.gradle:4]: extra space before `Versions`

**Review by BradleyWire — CHANGES_REQUESTED** [review]

File: zmessaging/build.gradle (hunk 7 above)
Comment (mejdoo) [reply zmessaging/build.gradle:58]: Good catch! I Will keep only those repositories below `zmessaging`

```
 maven { url "https://oss.sonatype.org/content/repositories/releases" }
 maven { url "https://oss.sonatype.org/content/repositories/snapshots" }
```

The rest are already fetched in `allprojects`

File: build.gradle (hunk 9 above)
Comment (mejdoo) [reply build.gradle:50]: There is a `quality.gradle`  file below `app/config` containing lint configuration. I will merge them all in the next PR.

File: scripts/customization.gradle (hunk 10 above)
Comment (mejdoo) [reply scripts/customization.gradle:1]: All the code here is about custom configuration. What do you suggest as name?

File: scripts/ci.gradle (hunk 6 above)
Comment (gizemb) [reply scripts/ci.gradle:119]: i don't know. maybe directly into build.gradle? it's not very important though, it can stay here as well

**Review by gizemb — APPROVED** [review]

File: scripts/customization.gradle (hunk 10 above)
Comment (BradleyWire) [reply scripts/customization.gradle:1]: custombuild maybe?

**Review by BradleyWire — APPROVED** [review]

**Comment by mejdoo** [comment]
build again

File: scripts/customization.gradle (hunk 10 above)
Comment (mejdoo) [reply scripts/customization.gradle:1]: I will rename it to `whitelabel`. I think we use this term at wire.

https://wearezeta.atlassian.net/wiki/spaces/PROD/pages/1704675/Custom+versions+and+whitelabel

**Pull request merged**